--- gnome-shell-3.24.1/js/misc/util.js~	2017-03-23 19:20:46.000000000 +0000
+++ gnome-shell-3.24.1/js/misc/util.js	2017-04-29 22:27:12.184426610 +0000
@@ -519,7 +519,7 @@
         if (schema) {
             this._setSettings(new Gio.Settings({ settings_schema: schema }));
         } else if (this._app) {
-            Mainloop.timeout_add_seconds(1, () => {
+            Mainloop.timeout_add_seconds(5, () => {
                 this._checkSettings();
                 return GLib.SOURCE_REMOVE;
             });
